### ISSUE TITLE
Help with installing clink-completions

### ISSUE THREAD

**Issue by eblis** [chrisant996/clink#324]
Sorry to bother you but it seems like I'm too thick to correctly install clink-completions and get them to work :(

I downloaded the latest release of [clink completions](https://github.com/vladimir-kotikov/clink-completions/releases/tag/0.3.7), put it in `%LOCALAPPDATA%/clink/completions` folder but when I try to use a git completion something crashes.
Immediately after I type in `git` I get this error:
![image](https://user-images.githubusercontent.com/540188/178692080-ca5a19bd-5c7d-42bb-b2b8-aa3cd06fd71c.png)

Looks like I'm missing some `path` library, but I'm not sure how and where to install it.

**Comment by eblis** [comment]
The `clink-completions` "release" has a modules folder which includes `path.lua` but it's not being picked up.
I tried adding the `modules` folder as a scripts path using `clink installscripts %LOCALAPPDATA%\clink\completions\modules\` but that only made things crash worse.
I now get this when clink starts.
![image](https://user-images.githubusercontent.com/540188/178694740-af64b1c3-c2ec-4cf5-ab48-07272da1d087.png)

All the files it mentions are available in the same script folder, but they're not being found for some reason.

**Comment by eblis** [comment]
So I managed to get of the errors by actually copying the module\*.lua files where the error complained they couldn't be found, more specifically `c:\windows\system32\lua` folder and now the error is no longer shown when I write the git command.

While suggestions are available for the `git` command, if I ctrl+space i get the list of commands, I'm not seeing the branch names for `git branch` command for example.
I really hoped the completions would help with the branch names as well

**Comment by eblis** [comment]
I have the following file in `%LOCALAPPDATA%\clink\completions` `.init.lua`:
```
-- The line below extends package.path with modules
-- directory to allow to require them
package.path = debug.getinfo(1, "S").source:match[[^@?(.*[\/])[^\/]-$]] .."modules/?.lua;".. package.path
```
This is the first part of the directory, I even extracted the .vscode folder :)
![image](https://user-images.githubusercontent.com/540188/178774239-7eaec06e-05b6-4b5f-a43d-977fe13c7d1c.png)

**Comment by eblis** [comment]
I installed clink over version 0.4.9, I didn't uninstall the old version first.
Could it be because of that ?

I also had some powerline for clink stuff (for 0.4.9), but I've since replaced those with [clink-flex-prompt](https://github.com/chrisant996/clink-flex-prompt)

**Comment by chrisant996** [comment]
Oh, thanks for sharing the pathname where clink-completions was extracted.

You can't extract clink-completions to a `completions\` directory under a directory that is already a script directory.

The `completions\` directory is special.  It's for delay-loading argmatcher scripts.  Like how the bash, fish, zsh, etc shells do.
The reason the clink-completions scripts are getting loaded at all is just an accident due to them being in that directory.

Scripts in a `completions\` directory are only loaded if the script name is typed.  E.g. you typed `git`, so `git.lua` got loaded -- but you haven't yet typed `.init`, so the `.init.lua` script hasn't been loaded yet, so none of the other clink-completions scripts will work.

They need to go in either the profile directory itself, or in a separate directory (not named "completions") and then use `clink installscripts dirname` or add the path to `set CLINK_PATH=...` or to `clink set clink.path ...`.

**Comment by eblis** [comment]
Sorry, I didn't know I couldn't use `completions` folder.

I renamed the folder to `extra` and used `clink installscripts <new folder path>`
![image](https://user-images.githubusercontent.com/540188/178780525-d70a4d53-79c9-4b04-8116-9ef72bf8f29a.png)

Now I don't get any errors but I don't think the completions work as expected.
I browsed to a git repository and typed `git switch` and then ctrl + space but i only got suggestions for files, not branches.

**Comment by eblis** [comment]
Looking in `clink.log` i have the following:
```
5c6c start_logger              157 ---- 2022/07/13 19:10:14.015 -------------------------------------------------
5c6c start_logger              163 Host process is 'cmd.exe' (pid 23660)
5c6c start_logger              167 DLL path is 'C:\Program Files (x86)\clink'
5c6c start_logger              181 Windows version 10.0.19044 (x64)
5c6c start_logger              184 Clink version 1.3.36.32d0bc (x64)
5c6c host_cmd::validate        304 Host is not interactive; cancelling inject.
```

But if I try to `clink inject` it says `Clink already loaded in process 37776`
I'm using Windows terminal if that makes a difference

**Comment by eblis** [comment]
I have another question though and I don't really want to open a new issue just for that.

The first time I type an invalid command (for example `qasdfalkhdgf`) in the command prompt and wait more than a few ms (so it doesn't happen immediately, but after I stop typing for a few ms or I type space or backspace) the process hangs for a while.
Have you experienced anything similar ?

I suspect it may have something to do with my shitty antivirus, but am not sure. Does clink search for the command once I start typing, does it try to execute it in another thread maybe ?
I know that my shitty antivirus will hang the command prompt for a long time if I actually try to run a command that doesn't exist in cmd prompt.

**Comment by chrisant996** [comment]
> ```
> 5c6c start_logger              157 ---- 2022/07/13 19:10:14.015 -------------------------------------------------
> 5c6c start_logger              163 Host process is 'cmd.exe' (pid 23660)
> 5c6c start_logger              167 DLL path is 'C:\Program Files (x86)\clink'
> 5c6c start_logger              181 Windows version 10.0.19044 (x64)
> 5c6c start_logger              184 Clink version 1.3.36.32d0bc (x64)
> 5c6c host_cmd::validate        304 Host is not interactive; cancelling inject.
> ```
> 
> But if I try to `clink inject` it says `Clink already loaded in process 37776` I'm using Windows terminal if that makes a difference

You installed Clink for autorun.  So, every Cmd invocation will autorun `clink inject`.  So, it must avoid actually initializing Clink if a given instance is not interactive (to improve performance, and to minimize potential for interference).  And, since Clink is already automatically injected, manually using `clink inject` will report that Clink is already loaded (because it is).

Every time any Clink instance starts, it tries to clear the log.  So, when configured for autorun, every Cmd will try to clear the log.  There are tons of Cmd.exe invocations under the covers by programs, so a lot more Cmd.exe invocations are happening than one might guess.

**Comment by chrisant996** [comment]
Oh, also -- does it only hang the _first_ time you type an unrecognized program name?  Or does it happen each time you type a new, different unrecognized program name in a Clink session?

(It remembers the result from each program name, so typing the same program name can simply reuse the previous results -- so typing the same program name multiple times would be very unlikely to hang.)

**Comment by eblis** [comment]
Sorry, I didn't want to pollute the issue tracker with something that might be related to my antivirus.
But I can create a new issue if you want.

The hang happens multiple times, but there's a "calm period" between the hangs, it won't happen immediately after another hang even if the command I've typed is still wrong. But after a while it will hang again.
I tried to have the same wrong name typed as in the previous hang but it's hard to time it and I don't have any conclusions about this yet (if caching of wrong results make clink not hang twice for the same wrong command
I'm using Process Hacker so there's a bit more info I can grab for the process(es), if you it.

[cmd.14812.zip](https://github.com/chrisant996/clink/files/9108376/cmd.14812.zip)
